Allow mod-tap hold action on one shot layer

kosorin opened the pull request:
## Description

This proposal allows mod-tap hold action on one shot layer. One shot layer is immediately disabled if `MT()` is pressed and therefore mod-tap hold is not possible.

### Example:

Following example currently produce a lower case character `x`.

```
 [L1] = LAYOUT(KC_X, OSL(L2))
 [L2] = LAYOUT(KC_A, LSFT_T(...))
```

- on layer 1
- `OSL(L2)` down
- `OSL(L2)` up
- on layer 2
- `LSFT_T(...)` down
- `KC_A` down
- `KC_A` up
- `LSFT_T(...)` up


With the proposed change the result is a upper case `A` and it will behave same as `KC_LSFT`.


I'm not aware if this covers all possible situations, but I've tested it for several months with no problems.


## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

drashna commented on quantum/action.c:
```diff
@@ -356,7 +356,16 @@ void process_action(keyrecord_t *record, action_t action) {
     bool do_release_oneshot = false;
     // notice we only clear the one shot layer if the pressed key is not a modifier.
-    if (is_oneshot_layer_active() && event.pressed && (action.kind.id == ACT_USAGE || !IS_MOD(action.key.code))
+    if (is_oneshot_layer_active() 
+        && event.pressed 
+        && (action.kind.id == ACT_USAGE
+            || !(IS_MOD(action.key.code)
+#    ifndef NO_ACTION_TAPPING
+                || action.kind.id == ACT_LMODS_TAP
+                || action.kind.id == ACT_RMODS_TAP
+#    endif
+                )
+            )
```
Could you reformat that (`qmk format-c`)?

drashna reviewed (changes requested):
Also, could you add to the unit tests to ensure that this behavior is properly tested?

kosorin commented:
I added a few unit tests and fixed a small bug that I discovered thanks to the tests. I'm not sure that's enough. It certainly doesn't test all permutations of all Tap-Hold options.

drashna commented:
> I added a few unit tests and fixed a small bug that I discovered thanks to the tests. I'm not sure that's enough. It certainly doesn't test all permutations of all Tap-Hold options.

Well, having them at all is a huge thing.  And they've already proven when even basic tests are definitely better than nothing, here. :)

At the very least, it helps with confidence in the code. And more robust testing can be added later.

KarlK90 reviewed (approved):
Thanks for fixing bugs and adding unit-tests in the progress! This is much appreciated. The changes LGTM.
